app/controllers/api/v1/media_controller.rb

```diff
@@ -122,53 +92,66 @@ def controlled_scope
 
   private
 
-  def precondition_fails?
-    query = if association_numeration == :single
-              media
-            else
-              Medium.where(id: params[:id])
-            end
-    run_etag_validation(query)
-  end
-
   def resource_scope(resources)
     return resources if resources.is_a?(ActiveRecord::Relation)
     Medium.where(id: resources.try(:id) || resources.map(&:id))
   end
 
+  def association_reflection
+    resource_class.reflect_on_association(media_name)
+  end
+
   def association_numeration
-    assoc = resource_class.reflect_on_association(media_name)
-    case assoc.macro
+    case association_reflection.macro
     when :has_one
       :single
     when :has_many
       :collection
     end
   end
 
-  def media_exists?
-    return false unless media
-    case association_numeration
-    when :single
-      media.exists?
-    when :collection
-      media.exists?(params[:id])
+  # store the original parental contolling scope wired up via the
+  # resource params and the resource_name method
+  # E.g project_id: 1 => Project is the controlled resource
+  def media_parental_controlled_resources
+    @media_parental_controlled_resources ||= controlled_resources
+  end
+
+  # override the controlled resources to handle the polymorphic media lookup
+  # from the different routes paths, e.g. /api/workflows/:id/attached_images
+  # will lookup the linked attached_images media for the workflow :id
+  def media_controlled_resources
+    linked_media_scope = association_reflection.klass.where(
+      linked_id: media_parental_controlled_resources.select(:id),
+      linked_type: resource_class.name
+    )
+    if params.key?(:id)
+      linked_media_scope = linked_media_scope.where(id: params[:id])
+    end
+    @controlled_resources = linked_media_scope
```

Comment:
Instead of unsetting memoized instance variables here try and provide a hook to `controlled_resources` behaviour., e.g. 
``` ruby
def controlled_resources
  find_controlled_resources(resource_class, resource_ids)
end

def find_controlled_resources(controlled_class, controlled_ids)
  @controlled_resources ||= api_user.do(controlled_scope)
     .to(resource_class, scope_context, add_active_scope: add_active_resources_scope)
     .with_ids(resource_ids)
     .scope
end
```
around here and in the `media_controller`
```
def controlled_resources
  find_controlled_resources(polymorphic_class, polymorphic_resource_ids)
end
```
https://github.com/zooniverse/Panoptes/blob/e33ae3cac8f67f646be0033aee60833458f22702/lib/role_control/roled_controller.rb#L22

thanks to @marten